### ISSUE TITLE
Update TFS Ban System

### ISSUE THREAD

**Issue by vankk** [otland/forgottenserver#2239]
I think that is time to we update our currently banishment system, that is too old.

We should a few things that had in TFS 0.4 for example:
- Notations;
- Bans by hours and not only just for days
- Bans should leaves at the time that the people with access added and not need wait until the next server save to move to another database and then the players can join the world again.

What do you guys think?

Regards,
vankk.

**Comment by gugahoa** [comment]
It's always good to have a nicer control over ban duration.

**Comment by vankk** [comment]
@djarek @ranisalt I'm interessed in doing that, how you guys want it?

**Comment by ranisalt** [comment]
IMHO it should be mostly moved to Lua and very configurable, with the defaults being what Tibia does today.

- You should ban by seconds (e.g. `player:ban(7 * 24 * 60 * 60)` to ban for a week)
- It should work with events (e.g. `onKill`/`onDeath` to record a frag)

I will assign it to you, but please open a PR as soon as you have some code and mark as [wip] so we can track progress :)

**Comment by vankk** [comment]
I was thinking in:

- We don't need the player to be online anymore. The Gamemaster can apply the ban even with the player offline. 
- What param we should use? /ban target, reason, type (e.g. ```/ban Life, suck, 30h``` that will ban the Life for 30 hours and we can use ```/ban Life, suck, 30d``` too, to ban the Life for 30 days. If you have a better suggestion for param please say it.

What you think about? Let me know if you have something else in mind.

Also, I didn't catch what you said about the events, you mean frag punishiment? If yes, how that should be done, in config.lua?

I'll be free in a couple days, and I'll start it, 😄

**Comment by ranisalt** [comment]
Isn't there a ban wizard embedded in the client?

**Comment by wgrriffel** [comment]
@ranisalt used to be, but it was removed as soon as gamemasters get removed

**Comment by yrpen** [comment]
Maybe create lua function and after that create script using modal windows?

**Comment by vankk** [comment]
I should proceed with the params @ranisalt ? That should be our best option for this since we didn't have the ban wizard anymore.

**Comment by ranisalt** [comment]
Creating a wizard with modal windows is a good one. Wait for more suggestions.

**Comment by vankk** [comment]
How that should be in modal window? We can't write on it. Also, if we proceed with modal windows we must approve to this repository the Modal Helper maded by @socket2810 .

**Comment by joseluis2g** [comment]
Not really, offline trainer windows is a perfect example.

**Comment by wgrriffel** [comment]
@joseluis2g what he meant is that modal window doesn't accept user input, imo modal window would look ugly, i think we should support classic banishment by chat commands, for a ban wizard it could be a patch for otc.

**Comment by Kamenuvol** [comment]
We need a Player(id) function to temporary load offline player. So player:ban could be used with them.

**Comment by ranisalt** [comment]
But the house list modal is writable

**Comment by wgrriffel** [comment]
@ranisalt that is not a modal window, it's a client built-in window dialog, same for books/letter editing, these allow you to get user input, that why i think it would be ugly.

**Comment by vankk** [comment]
I agree with @wgrriffel . We should proceed with params, something in mind? I think the way that I [suggested](https://github.com/otland/forgottenserver/issues/2239#issuecomment-299619277) is awesome!

**Comment by wgrriffel** [comment]
@vankk I think you should make it according with new report system, /ban target, type, reason, comment, but not type as time, as a type itself, each type has a ban time, reason could be reasons that has in report window, comment could be a custom one by administrator or automated to get type + reason, however, to it works means that we should support new reporting things, i'll work on it.

**Comment by securmk** [comment]
I'll do some research on this.

**Comment by M4GO** [comment]
`local config = {
    sep = '\n\n----------------------------------------------------------\n',
    file = "data/logs/bans.txt"
}
function onSay(cid, words, param)
	local file = io.open(config.file, "a+")
	if not file then
	return print("Warning: Could not open ".. config.file)
	end
    local player = Player(cid)
    if player:getAccountType() < ACCOUNT_TYPE_GOD then
        return false
    end
 
    local split = param:split(",")
    if split[1] ~= nil then
        local accountId = getAccountNumberByPlayerName(split[1])
        if accountId > 0 then
            local comment = ""
            if split[2] == nil then
                player:sendTextMessage(MESSAGE_EVENT_ADVANCE, "Usage: /ban Nick, days, reason\nFor example: /ban PlayerName, 1, Bug abuse")
                return false
            elseif isNumber(split[2]) == false then
                player:sendTextMessage(MESSAGE_EVENT_ADVANCE, "Usage: /ban Nick, days, reason\nFor example: /ban PlayerName, 1, Bug abuse")
                return false
            end
            if split[3] == nil then
			 player:sendTextMessage(MESSAGE_EVENT_ADVANCE, "Usage: /ban Nick, days, reason\nFor example: /ban PlayerName, 1, Bug abuse")
             comment = split[3]
                return false
            end
        else
            player:sendCancelMessage("Player with name " .. split[1] .. " doesn't exist.")
            return false
        end
     
        local resultId = db.storeQuery("SELECT 1 FROM `account_bans` WHERE `account_id` = " .. accountId)
        if resultId ~= false then
            result.free(resultId)
            return false
        end
     
        local targetCid = getPlayerByName(split[1])
        local timeNow = os.time()
        local queryBool = db:query("INSERT INTO `account_bans` (`account_id`, `reason`, `banned_at`, `expires_at`, `banned_by`) VALUES (" .. accountId .. "," .. "'" .. tostring(split[3]) .. "'" ..", " .. timeNow .. ", " .. timeNow + (split[2] * 86400) .. ", " .. getPlayerGUIDByName(getCreatureName(cid)) .. ")")
 
        if queryBool == true then
            if targetCid ~= false then
                doRemoveCreature(targetCid)
            end
       player:sendTextMessage(MESSAGE_EVENT_ADVANCE,"" .. split[1] .. " was banished because of " .. split[3] .. " for " .. split[2] .. " days by " .. player:getName() .. ".")
			file:write(string.format('Date [%s] |%s Banished by %s because of%s%s', os.date("%d/%m/%y %H:%M"),  split[1], player:getName(), split[3], config.sep))
			file:close()
        else
            player:sendTextMessage(MESSAGE_EVENT_ADVANCE, "Usage: /ban Nick, days, reason\nFor example: /ban PlayerName, 1, Bug abuse")
        end
    end  
end

`

**Comment by Codinablack** [comment]
Would it be going too far to make the updated ban system a sophisticated Modal Window?

**Comment by DSpeichert** [comment]
Also related to #2178 for IPv6 support in bans.
